syntax = "proto3";

// NEXT TAG = 9
message FrameU8 {
  string field_name_1 = 1;
  string field_name_2 = 2;
  string field_name_3 = 3;
  string field_name_4 = 4;
  string field_name_5 = 5;
  string field_name_6 = 6;
  string field_name_7 = 7;
  string field_name_8 = 8;
}

// NEXT TAG = 5
message FrameU16 {
  string field_name_1 = 1;
  string field_name_2 = 2;
  string field_name_3 = 3;
  string field_name_4 = 4;
}

// NEXT TAG = 3
message FrameU32 {
  string field_name_1 = 1;
  string field_name_2 = 2;
}

// NEXT TAG = 2
message FrameU64 {
  string field_name_1 = 1; //
}

// NEXT TAG = 1
message FrameEmpty {}

// NEXT TAG = 7
message CanData {
  uint64 data = 1;
  oneof frame {
    FrameU8 u8 = 2;
    FrameU16 u16 = 3;
    FrameU32 u32 = 4;
    FrameU64 u64 = 5;
    FrameEmpty empty = 6;
  }
}

// NEXT TAG = 9
message CanMsg {
  enum Source {
    PLUTUS = 0;
    CHAOS = 1;
    TELEMETRY = 2;
    LIGHTS_FRONT = 3;
    LIGHTS_REAR = 4;
    MOTOR_CONTROLLER = 5;
    DRIVER_CONTROLS = 6;
    DRIVER_DISPLAY = 7;
    SOLAR_MASTER_FRONT = 8;
    SOLAR_MASTER_REAR = 9;
    SENSOR_BOARD = 10;
  }

  uint32 raw_id = 1; // 0-2047 only
  Source source = 2; // 0-15 only
  bool is_ack = 3;
  uint32 id = 4; // 0-63 only
  CanData can_data = 5;
  string msg_name = 6;
  string msg_readable_name = 7;
  bool is_critical = 8;
}

// NEXT TAG = 2
message CanSchema {
  repeated CanMsg msg = 1; //
}
